File: StatusBarDemo/app/src/main/java/com/renyu/statusbardemo/activity/BaseActivity.java
package com.renyu.statusbardemo.activity;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import static com.renyu.statusbardemo.utils.BarUtils.setColor;
import static com.renyu.statusbardemo.utils.BarUtils.setTranslucent;

/**
 * Created by renyu on 2017/4/11.
 */

public abstract class BaseActivity extends AppCompatActivity {

    public abstract void initParams();
    public abstract int initViews();
    public abstract int setStatusBarColor();
    public abstract int setStatusBarTranslucent();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (initViews()!=0) {
            setContentView(initViews());
        }

        // 设置沉浸式，二选一
        if (setStatusBarColor()!=0) {
            setColor(this, setStatusBarColor());
        }
        if (setStatusBarTranslucent()!=0) {
            setTranslucent(this);
        }

        initParams();
    }
}

File: ArchitectureDemo/app/src/main/java/com/renyu/architecturedemo/mvvm/base/BaseActivity.kt
package com.renyu.architecturedemo.mvvm.base

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v7.app.AppCompatActivity

abstract class BaseActivity<T : ViewDataBinding, M : Any, VM : Any> : AppCompatActivity() {

    abstract fun initViews(): Int
    abstract fun variableId(): Int
    abstract fun model(): M
    abstract fun viewModel() : VM
    abstract fun initParams()
    abstract fun loadData()

    var viewModel: VM? = null

    var viewDataBinding: T? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewDataBinding = DataBindingUtil.setContentView<T>(this, initViews()).also {
            it.setVariable(variableId(), model())
        }

        viewModel = viewModel()

        initParams()
        loadData()
    }
}